### PR TITLE
update visualizer to correctly do charts for all runs

### PR DESCRIPTION
**Pull request by amj**
some pretty thorny code in here.

### CODE REVIEW

File: oneoffs/joseki.html [hunk 1]
```diff
@@ -56,40 +49,41 @@ <h1> Minigo Opening Analysis </h1>
...
-                      var data = new google.visualization.DataTable();
-                      data.addColumn('number', 'Timestamp'); // Implicit domain label col.
-                      data.addColumn('number', 'Frequency'); // Implicit series 1 data col.
-                      data.addRows(this.top10[index].chart);
+                      var data = new google.visualization.DataTable({
+                          cols: this.topSeq[index].cols,
+                          rows: this.topSeq[index].rows},
+                      0.6);
+
+                      //data.addColumn('number', 'Timestamp'); // Implicit domain label col.
```
Comment (sethtroisi) [inline oneoffs/joseki.html]: cleanup?

File: oneoffs/opening_freqs_export.py [hunk 2]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
 """
 
 
+def get_runs(db):
+    return [r[0] for r in db.execute('''
+               select distinct(run) from joseki_counts group by 1'''
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: does this really require "group by 1"?

File: oneoffs/opening_freqs_export.py [hunk 3]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+def top_seqs_by_run(db, topN=300):
+    runs = get_runs(db)
+    top_seqs_by_run = {}
+    for r in runs:
+        cur = db.execute('''
+                         select seq, sum(count) from joseki_counts
+                         where run = ?
+                         group by seq order by 2 desc limit ?;
+                         ''', (r, topN))
+        sequences = [row for row in cur.fetchall()]
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: list(cur.fetchall())

File: oneoffs/opening_freqs_export.py [hunk 4]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+        top_seqs_by_run[r] = sequences
+    return top_seqs_by_run
+
+
+def run_time_ranges(db):
+    ts = lambda hr: int(dt.datetime.strptime(hr, "%Y-%m-%d-%H").timestamp())
+    runs = [ (r[0], (ts(r[1]), ts(r[2]))) for r in db.execute('''
+        select run, min(hour), max(hour) from joseki_counts group by 1;
+        ''').fetchall()]
+    return dict(runs)
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: {r[0]: (ts(r[1]), ts(r[2])) ....}

or

{r: (ts(min_t), ts(max_t)) for r, (min_t, max_t) in db.execute...

File: oneoffs/opening_freqs_export.py [hunk 5]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+
+
+def build_run_time_transformers(ranges):
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: ranges, buckets=250)

File: oneoffs/opening_freqs_export.py [hunk 6]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
-        obj = {"count": count, "chart": dat, "sequence": seq}
-        data.append(obj)
-
-    print('saving')
-    tmpl = Template(open('oneoffs/joseki.html').read())
-    with open(FLAGS.out_file, 'w') as out:
-        out.write(tmpl.render(giant_blob=json.dumps(data)))
+    ranges = run_time_ranges(db)
+    interps = build_run_time_transformers(ranges)
+    seqs_by_run = top_seqs_by_run(db, 100)
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: top_n=100

File: oneoffs/opening_freqs_export.py [hunk 7]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+    for run in runs:
+        data = []
+        sequences = seqs_by_run[run]
+        for seq, count in sequences:
+            print(run, seq, count)
+            rows = collections.defaultdict(lambda: [0 for i in range(len(runs))])
+
+            for idx, r in enumerate(runs):
+                cur = db.execute('''
+                                 SELECT hour, count, run from joseki_counts where seq = ? and run = ?;
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: silly to return run when you know what is it.

SELECT hour, count from joseki_counts where seq = ? and run = ?;

File: oneoffs/opening_freqs_export.py [hunk 8]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+                                 ''', (seq, r))
+
+                dat = cur.fetchall()
+                for d in dat:
+                    key = interps[d[2]](ts(d[0]))
+                    rows[key][idx] = d[1]
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: for hour, count in cur.fetchall():
    key = interps\[run](ts(hour))
    rows[key][idx] = count

File: oneoffs/opening_freqs_export.py [hunk 9]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+
+                #dat = [ {'c': [ {'v': interps[d[2]](ts(d[0]))} ] + onehot(d[1], idx, len(runs))
+                #       } for d in dat]
+                #rows.extend(dat)
+            row_data = [ {'c': [ {'v': key} ] + [{'v': v if v else None} for v in value ] } for key,value in rows.items()]
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: line break before for key, value

File: oneoffs/opening_freqs_export.py [hunk 10]
```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
...
+            obj = {'run': run, "count": count, 'cols': cols, "rows": row_data, "sequence": seq}
+            data.append(obj)
+
+        print('saving')
+        tmpl = Template(open('oneoffs/joseki.html').read())
+        with open(run + FLAGS.out_file, 'w') as out:
+            out.write(tmpl.render(giant_blob=json.dumps(data), run=run, time_ranges=json.dumps(ranges)))
+
+def onehot(v, idx, max_):
```
Comment (sethtroisi) [inline oneoffs/opening_freqs_export.py]: Not needed anymore?

Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 9 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 8 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 7 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 6 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done and added a flag.

File: oneoffs/opening_freqs_export.py (hunk 5 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 4 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/opening_freqs_export.py (hunk 3 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done and below.

File: oneoffs/opening_freqs_export.py (hunk 2 above)
Comment (amj) [reply oneoffs/opening_freqs_export.py]: Done.

File: oneoffs/joseki.html (hunk 1 above)
Comment (amj) [reply oneoffs/joseki.html]: Done.

**Pull request merged**